README.md
# scs_model
Code for computational modeling of spinal cord stimulation to evoke lower urinary tract nerve activity and bladder pressures.

## STEPS: 

## Step 1: Image segmentation and tractography

### 1.1 Automatic segmentation
This step is run from the code '****' and performs automatic segmentation of the spinal cord tissues (gray matter, white matter, dorsal/ventral roots and rootlets, cerebrospinal fluid, and epidural fat) in the MR dataset using a pre-trained Deep Neural Network. It generates and saves the segmentation masks, Jaccard Index and Hausdorff Distance metrics into the output folder. This code may be run with any python interpreter. Inspection and correction of the segmentation can be performed with any image processing software, such as iSeg or FSL.

**Input:** MR dataset (.nii)

**Output:** Automatically segmented images (.nii), Jaccard index & Hausdorff distance (.txt)

**System Requirements:**
need to add

**Instructions:** The user needs to upload a zipped directory containing the following folders: 
1. The user should create an **"input"** folder that contains the acquired imaging data with file with names **19f, t2** and **fat**. These files can either be in the .nii or the .nii.gz format. Note that the input dimension of these images is fixed (see below).

2. The user should crease an **"output"** folder that contains a few manually segmented examples for each tissue to be segmented. This is to allow masking of the spinal cord for better performance. The files should have the following names:
csf.nii.gz, dorsalroots.nii.gz, epiduralfat.nii.gz, graymatter.nii.gz, ventralroots.nii.gz, whitematter.nii.gz

3. Run the code '****' from the folder containing the "input" and "output" folders.

4. Check the segmentation results.

The default Neural Network (hybrid cat1-cat3) used in this node was trained with manual segmentations of 2 cat lumbosacral spine samples distributed along the cord (cat spine 1: 207 slices & cat spine 3: 176 slices). Note that to use this network, the raw images have to be of 384 x 512 pixel dimension in the axial plane.

![Asset 3](https://user-images.githubusercontent.com/43448105/191312158-00045d2b-d32c-4bdf-8836-307683e137f6.png)

**Limitations**: This pipeline currently does not support training new weights for the Neural Network. If the user has data that is substantially different from the cat lumbosacral spine (ex. cat cervical spine) and simply running the default does not give satisfactory segmentations, the user may upload a Neural Network pre-trained with manual segmentation of the user’s data. To generate this pre-trained Network in the correct .pickle format, contact Alessandro Fasse (fasse@zmt.swiss).

### 1.2 Generate neuron axon trajectories
This step does not require a code, but is necessary to perform to generate neuron trajectories for functionalizing the spinal cord model. We suggest using [DSI_Studio](https://dsi-studio.labsolver.org/), a tractography software tool for diffusion MRI processing. It allows construction of realistic axonal trajectories based on DTI data, which can later be placed as neurons in the 3D model.

**Input:** 1-DTI images (.nii), 2-b-table (.txt)
**Output:** tracked fiber coordinates (.txt)

**Instructions:** 
1. In Step T1, import DTI image files. You will be prompted to add the b-table.
2. In Step T2, import the .src file generated in Step T1, select DTI checkbox, Check b-table checkbox. Click on Run Reconstruction.
3. In Step T3, import the .fib file generated in Step T2.
4. Tracking:
*Region of Interest (ROI):*
- For each root level, create a new region of interest by selecting Regions -> New region.  
- Draw the region of interest by selecting a desired image slice and drawing around the spinal root. We suggest picking a slice just rostral to the DRG of a target root. Please note that for the default use of the pipeline, the fibers should be tracked individually for each root level (L6, L7 etc.), side(ventral/dorsal) and location (right/left).
**Example:** for spinal roots L6 the user will have to track and save separately the following files: L6_r_d (right, dorsal), L6_l_d (left,dorsal), L6_r_v (right, ventral), L6_l_v (left, ventral).

*Tracking parameters:*
Tracking parameters may differ depending on the length of your specific dataset. We suggest the following values for a 8cm long dataset: Angular Threshold=60, Min Length=20, Max Length=100, Terminate if=10000 Tracts.
Click on Fiber Tracking and visualize the result.

*Tract cleaning:*
To ensure that most tracts generated are continuous from root into the spinal cord, cleaning is suggested. Find the level where the tracts enter the spinal cord, press Ctrl+S and draw a line across the spinal cord. This will ensure only tracts that pass that line will be kept.

![Asset 8](https://user-images.githubusercontent.com/43448105/191318832-e0b5d978-e7cb-4d73-b1a8-889e8b4931b9.png)

5. To save tracts, click on Tracts and save created tracts in the Output folder as a .txt file, named as previously described.
6. Repeat for each root level, side and location.

## Step 2: Simulate neural recruitment

This step is run from the code 'run_complex_model.py' and includes building a finite element model (FEM) based on the imaging data provided, simulates the electromagnetic potential evoked in that model during simulation, populates the model with neural trajectories, and simulates actual neural recruitment on those neurons. All of the neural recruitment data is then saved to an output file. To run this stage, open run_complex_model.py in a version of Sim4Life 7 or higher that includes A-delta neuron modeling capabilities, and in the "runProject" function change template and contact_n to be the name of the electrode file to load and the contact number to use, respectively. Change num_neuron to be the number of neurons per root and type with which to run the simulation. Additionally, pathnames in main() will need to be changed to local directories. 

After results are saved from the neuron simulation, run plot_analysis_scale.ipynb to re-scale these values to the correct ratio of neurons per root and view initial plots of recruitment values for a simulation. This step will also output a series of recruitment curve values in a text file that can be used in Step 3. 

## Step 3: Simulate bladder function

Step 3 includes a spiking neural network model that receives inputs generaged from upstream FEM model for experiment validation and exploration. 

The model needs [NEURON](https://www.neuron.yale.edu/neuron/) installed to perform normally. Please follow the *README.ipynb* notebook to compile the environment and preprocess the data before running a simulation.

We provide two options for running a simulation: 

1. use computed recruitment_threshold_percentage (only pudendal afferents are recruited and simulated) to validate bladder behavior under specific pudendal afferent recruitment

2. use recruitment curve data (pudendal afferent, pelvic afferent and SPN are recruited at different stimulation amplitude (uA)) and explore the bladder behavior at any customized stimulation amplitude by defining "stim_amp_uA". 

For option 2, the "splitting" and "mapping" steps in *recruitment data splitter.ipynb* is necessary.

All results are stored in ../results. There is also an analysis script Analysis on bladder traces and neuron firings.ipynb to provide some simple processing of data. The details of results are listed in *README.ipynb*.




step_3/CHANGELOG.md
# Changelog

## [1.1.0] - 2021-10-01
### Changed
- update installation of jupyter-commons with "jupyter-minimal" optional requirements
- changed base image to jupyter-base 
## [1.0.7] - 2021-06-22
### Changed
- updates simcore service library to reduce cpu usage when extracting archives
## [1.0.6] - 2021-04-06
### Fixed
- fixed simcore.service.settings docker labels concerning RAM/CPU limits

## [1.0.4] - 2021-03-22
### Fixed
- fixed requirements

## [1.0.2] - 2021-03-22
### Changed
- updated simcore-sdk to 0.3.2

## [1.0.1] - 2020-03-09
### Fixed
- state_puller no longer causes out of memory errors causing the entire app to crash
- state puller will no longer allow the notebook to boot upon error when recovering the status
### Changed
- upgraded simcore-sdk and dependencies

## [1.0.0] - 2021-02-19
### Added
- jupyter notebook with octave and python kernels
- python kernel installs math and NEURON libraries (see [requirements.txt](requirements.txt))
- jupyterlab installs the following extensions:
  - [git extensions](https://github.com/jupyterlab/jupyterlab-git#readme)
  - [interactive matplotlib](https://github.com/matplotlib/ipympl#readme)


---
All notable changes to this service will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and the release numbers follow [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


<!-- Add links here -->

<!-- [1.0.0]:https://git.speag.com/oSparc/sparc-internal/-/tags/jneuron_v1.0.0  -->


<!-- HOW TO WRITE  THIS CHANGELOG

- Guiding Principles
  - Changelogs are for humans, not machines.
  - There should be an entry for every single version.
  - The same types of changes should be grouped.
  - Versions and sections should be linkable.
  - The latest version comes first.
  - The release date of each version is displayed.
  - Mention whether you follow Semantic Versioning.
  -
- Types of changes
  - Added for new features.
  - Changed for changes in existing functionality.
  - Deprecated for soon-to-be removed features.
  - Removed for now removed features.
  - Fixed for any bug fixes.
  - Security in case of vulnerabilities.

SEE https://keepachangelog.com/en/1.0.0/
-->


step_2/plotting_fn.py
'''plotting functions for analysis'''
import os
from statistics import median
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import numpy as np
import random
from helper_fn import *

def generateNDist(data, inc_reclocs, perc_neurons):
    '''INPUTS: data, as a pandas data structure that contains the columns 
    ['name', 'nerve', 'diameter', 'elec_dist', 'enc_tissue', 'rec_amp'], 
    where name includes root level, 'Left' or 'Right' and 'Dorsal' or 'Ventral' to ID
    the neurons. inc_reclocs, as a list of strings indicating location of 
    first recruitment that should be included in final output.
    OUTPUT: same data structure, but with only the rows matching the given
    recruitment tissue, subselected and assigned to nerves by size distribution, and
    with specific columns labeling root, side, and type of neuron'''
    #remove data that is recruited in other tissues
    print('Generate neuron distribution')
    tmp = data[data['enc_tissue'].isin(inc_reclocs)].copy()
    #NEURON DISTRIBUTIONS
    #Axons per nerve - ratio at each level (each level sums to 100); [pelvic, pudendal, sciatic] %
    v_ratio = {'L6': [0, 0, 100], 'L7': [0, 0, 100], 'S1': [
        7, 62, 31], 'S2': [66, 29, 5], 'S3': [13, 87, 0]}
    d_ratio = {'L6': [0, 0, 100], 'L7': [0, 0, 100], 'S1': [
        9, 22, 69], 'S2': [55, 37, 8], 'S3': [64, 36, 0]}
    root_names = list(v_ratio.keys())
    #Typical size distribution for each NERVE - in same order as the v_ratio and d_ratio arrays (pelvic, pudendal, sciatic)
    sizebins = range(1, 21)
    d_sizedist = [[0, 55.9, 26.4, 5.9, 5.9, 5.9, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0.6, 3.5, 7.3, 10.3, 12.8, 13.8, 13.1, 12.1, 10.2,
                    8, 4.8, 1.2, 0.9, 0.7, 0.3, 0.4, 0, 0, 0, 0],
                [0, 0, 0.2, 4.5, 8.9, 10.0, 7.9, 6.9, 4.9, 4.5, 5.2, 5.3, 5.3, 6.7, 7.4, 7.3, 6.6, 4.8, 2.5, 1.1]]
    v_sizedist = [[0, 41.7, 50.9, 7.4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 4.6, 4.9, 11.5, 4.6, 5.9, 17.5, 13.2,
                    9, 6.2, 8.2, 0.1, 4.4, 0.9, 4.6, 4.4, 0, 0],
                [0, 0, 0.2, 4.5, 8.9, 10.0, 7.9, 6.9, 4.9, 4.5, 5.2, 5.3, 5.3, 6.7, 7.4, 7.3, 6.6, 4.8, 2.5, 1.1]]

    #add labels based on name of each neuron
    spinal_lvl_idx = np.empty(len(tmp), dtype='<U2')

    for i in range(len(root_names)):
        spinal_lvl_idx[tmp['name'].str.contains(root_names[i])] = root_names[i]

    tmp['root'] = spinal_lvl_idx
    tmp['type'] = [['afferent', 'efferent'][i]
                   for i in tmp['name'].str.lower().str.contains('ventral')*1]
    tmp['side'] = [['right', 'left'][i] for i in tmp['name'].str.lower().str.contains('left')*1]

    #based on these numbers and the neuron distributions, how many axons should be assigned to each nerve?
    #count the number of ie dorsal L6 neurons, multiply it by distribution in ratios
    d_arr = [[] for x in root_names]
    v_arr = [[] for x in root_names]
    nerve_idx = [-1 for x in range(len(tmp))]
    tmp['nerve'] = nerve_idx
    for i in range(len(root_names)):
        #assign how many of each nerve, based on level - do 50% of total neurons
        aff_tmp = tmp[(tmp['root'] == root_names[i]) & (tmp['type'] == 'afferent')]
        eff_tmp = tmp[(tmp['root'] == root_names[i]) & (tmp['type'] == 'efferent')]
        ddist = [x*.0001*perc_neurons*len(aff_tmp) for x in d_ratio[root_names[i]]]
        vdist = [x*.0001*perc_neurons*len(eff_tmp) for x in v_ratio[root_names[i]]]

        #the only reason to index this like this is to be able to save the whole
        # expected distribution to compare
        d_arr[i] = [[] for x in range(len(ddist))]
        v_arr[i] = [[] for x in range(len(ddist))]
        for j in range(len(ddist)):
            d_arr[i][j] = np.random.choice(sizebins, int(ddist[j]), p=[
                x*.01 for x in d_sizedist[j]])
            v_arr[i][j] = np.random.choice(sizebins, int(vdist[j]), p=[
                x*.01 for x in v_sizedist[j]])

        used_idx = []
        missing_n = [[] for x in range(len(ddist))]
        #find diameters of neurons that are AFFERENT and in the appropriate root
        for j in range(len(ddist)):  # for each nerve
            # for each size that was randomly called for in dorsal dist
            for k in range(len(d_arr[i][j])):
                #find an appropriate neuron.
                if d_arr[i][j][k] in list(aff_tmp['diameter']):
                    #find indices that have not already been chosen
                    remain_idx = list(
                        set(aff_tmp.index[aff_tmp['diameter'] == d_arr[i][j][k]])-set(used_idx))
                    if len(remain_idx) == 0:
                        missing_n[j].append(d_arr[i][j][k])
                        print('No remaining neuron of size ' + str(d_arr[i][j][k]))
                        #for now, just print this and then don't do anything about it... need to figure out how common it is
                        continue
                    #pick a random index
                    n_idx = random.choice(remain_idx)
                    #add it to the list of chosen indices - should be fastest to just save idx in separate list
                    tmp['nerve'].loc[n_idx] = j
                    used_idx.append(n_idx)
                else:
                    missing_n[j].append(d_arr[i][j][k])
                    print('No neuron of size ' + str(d_arr[i][j][k]))

        #Do the same for neurons that are EFFERENT
        for j in range(len(vdist)):  # for each nerve
            # for each size that was randomly called for in vebtral dist
            for k in range(len(v_arr[i][j])):
                #find an appropriate neuron.
                if v_arr[i][j][k] in list(eff_tmp['diameter']):
                    #find indices that have not already been chosen
                    remain_idx = list(
                        set(eff_tmp.index[eff_tmp['diameter'] == v_arr[i][j][k]])-set(used_idx))
                    if len(remain_idx) == 0:
                        missing_n[j].append(v_arr[i][j][k])
                        print('No remaining neuron of size ' +
                              str(v_arr[i][j][k]))
                        #for now, just print this and then don't do anything about it... need to figure out how common it is
                        continue
                    #pick a random index
                    n_idx = random.choice(remain_idx)
                    #update list info 
                    tmp['nerve'].loc[n_idx] = j
                    used_idx.append(n_idx)
                else:
                    missing_n[j].append(v_arr[i][j][k])
                    print('No neuron of size ' + str(v_arr[i][j][k]))

    tmp['nerve'] = [['pelvic', 'pudendal', 'sciatic', 'undefined'][i] for i in tmp['nerve']]
    #TODO also output: d_sizedist, v_sizedist, d_arr, v_arr, sizebins
    expected_sizedist = {'sizes':sizebins, 'dorsal':d_sizedist, 'ventral':v_sizedist}
    actual_sizelst = {'dorsal':d_arr, 'ventral':v_arr}
    return tmp[tmp['nerve'] != 'undefined'], expected_sizedist, actual_sizelst




def plotNDist(ideal_dist, actual_dist, bins, savepath=''):
    '''input the ideal distribution as percent values for each of the bins, 
    input the actual distribution as the list of values assigned'''
    if savepath!='':
        f = open(savepath, 'w')
        lbl = savepath.split('neuronDist')[1]
    #print('this is new' + '2')
    allkeys = [x for x in ideal_dist.keys()]
    #print(actual_dist)
    binmid = [(x + y)/2+(y-x) for (x, y) in zip(bins[:-1], bins[1:])]
    for key in allkeys:
        # print(actual_dist[key])
        # print(ideal_dist[key])
        if len(actual_dist[key])==0:
            # print(key)
            continue
        #make a new plot
        fig, ax = plt.subplots()
        #plot actual distributions as a histogram
        plt.hist(actual_dist[key], bins, alpha = 0.6, color=[.24, .64, 1])
        plt.title(key)
        plt.xlabel('Axon size (um)')
        plt.ylabel('Count')

        #plot the ideal distribution as individual points, scaled for the number of neurons in actual dist
        vals = [x*len(actual_dist[key])/100 for x in ideal_dist[key]]
        # print(vals)
        # print(len(vals))
        # print(binmid)
        # print(len(binmid))
        plt.scatter(binmid, vals, marker = 'o', alpha = 0.8, c=[[0, 0, 0]])

        box_off(ax)
        imax = max([max(actual_dist[key]), max(vals)])
        plt.ylim([0, imax*1.2])
        reset_ticks(ax, [0, bins[-1]], [0, imax])

        if savepath!='':
            #save to file
            f.write(key + '\n')
            f.write('bins ' + str(binmid) + '\n')
            f.write('actual ' + str(actual_dist[key]) + '\n')
            f.write('ideal ' + str(vals) + '\n')

            figpath = os.path.join(savepath, '..', key.lower() + '_dist' + lbl[:-4] +'.png')
            plt.savefig(figpath)

def plotRootRec(df, figpath=''):
    '''input a filepath to save and a pandas dataframe containing afferents/efferents labeled
    in the "type" column and roots labeled in "root" column'''
    tmp1 = pd.melt(df, id_vars=['type'], value_vars=['root'])

    tmp2 = pd.melt(df, id_vars=['type'], value_vars=[
                'rec_amp'], value_name='Rec Amp', var_name='Root')
    tmp2['Root'] = tmp1['value']  # just have to massage this so labels are correct
    tmp2.sort_values(by='type') #this should keep the ordering same in all plots

    clr = ["#ffa800", "#a560bd"]
    newPal = dict(afferent=clr[1], efferent=clr[0])
    newPalDk = dict(afferent="#45005e", efferent="#a53d00")

    # Initialize the figure
    f, ax = plt.subplots()
    sns.despine(left=True)

    # Show each observation with a scatterplot
    sns.stripplot(x="Rec Amp", y="Root", hue="type",
                data=tmp2, dodge=True, alpha=.25, zorder=1, palette=newPal)

    # Show the conditional means, aligning each pointplot in the
    # center of the strips by adjusting the width allotted to each
    # category (.8 by default) by the number of hue levels
    sns.pointplot(x="Rec Amp", y="Root", hue="type",
                data=tmp2, dodge=.8 - .8 / 2,
                join=False, palette=newPalDk,
                markers="d", scale=.75, ci=None, estimator=median)

    # Improve the legend
    handles, labels = ax.get_legend_handles_labels()
    ax.legend(handles[2:], labels[2:], title="Type",
            handletextpad=0, columnspacing=1,
            loc="lower right", ncol=2, frameon=True)
    if figpath:
        plt.savefig(figpath)
